src/core/main.c

```diff
@@ -796,6 +816,80 @@ static void set_manager_settings(Manager *m) {
         m->status_unit_format = arg_status_unit_format;
 }
 
+static int set_cgroup_context_root_slice(Manager *m) {
+        Unit *u = NULL;
+        CGroupContext* cgroup_context;
+        _cleanup_io_cost_qos_free_ CGroupIOCostQos* io_cost_qos = NULL;
+        _cleanup_io_cost_model_free_ CGroupIOCostModel* io_cost_model = NULL;
+        char* io_cost_qos_path, *io_cost_model_path;
+
+        assert(m);
+
+        if ((u = manager_get_unit(m, SPECIAL_ROOT_SLICE))) {
+                if (!arg_io_cost_qos.path && !arg_io_cost_model.path)
+                        return 0;
+                if (arg_io_cost_qos.ctrl != IO_COST_CTRL_USER) {
+                        log_info("ctrl is not set to user for io_cost_qos, setting all other values to 0");
+                        arg_io_cost_qos.read_latency_percentile = 0;
+                        arg_io_cost_qos.read_latency_threshold = 0;
+                        arg_io_cost_qos.write_latency_percentile= 0;
+                        arg_io_cost_qos.write_latency_threshold= 0;
+                        arg_io_cost_qos.min= 0;
+                        arg_io_cost_qos.max= 0;
+                }
+                if (arg_io_cost_model.ctrl != IO_COST_CTRL_USER) {
+                        log_info("ctrl is not set to user for io_cost_model, setting all other values to 0");
+                        arg_io_cost_model.rbps = 0;
+                        arg_io_cost_model.rseqiops = 0;
+                        arg_io_cost_model.rrandiops = 0;
+                        arg_io_cost_model.wbps = 0;
+                        arg_io_cost_model.wseqiops = 0;
+                        arg_io_cost_model.wrandiops = 0;
+                }
+                cgroup_context = unit_get_cgroup_context(u);
+                if (arg_io_cost_qos.path) {
+                        io_cost_qos = new0(CGroupIOCostQos, 1);
+                        if (!io_cost_qos) {
+                                log_oom();
+                                return -ENOMEM;
+                        }
+                        memcpy(io_cost_qos, &arg_io_cost_qos, sizeof(CGroupIOCostQos));
+
+                        io_cost_qos_path = strdup(arg_io_cost_qos.path);
+                        if (!io_cost_qos_path) {
+                                log_oom();
+                                return -ENOMEM;
+                        }
+                        io_cost_qos->path = io_cost_qos_path;
+
+                        if (!cgroup_context->io_cost_qos)
+                                cgroup_context->io_cost_qos = TAKE_PTR(io_cost_qos);
+                }
+
+                if (arg_io_cost_model.path) {
+                        io_cost_model = new0(CGroupIOCostModel, 1);
+                        if (!io_cost_model) {
+                                log_oom();
+                                return -ENOMEM;
+                        }
+                        memcpy(io_cost_model, &arg_io_cost_model, sizeof(CGroupIOCostModel));
```

Comment:
why memcpy() if a plain assignment works too?

Reply:
so you first use new0() to allocate the struct, and the override it in full. So you filled the thing with zeroes for nothing?

Reply:
We decided against the plain assignment because we wanted the arg_io_cost_qos/model structs in main.c to be free'd separately from the io_cost_qos/model used in cgroup.c. In main.c they are free'd in `reset_arguments()` whereas in cgroup.c they are free'd in `cgroup_context_done()`.